### PR TITLE
Introduce treasury and document

### PR DESCRIPTION
**Pull request by gavofyork**
- Rename `Executable` to `OnFinalise` to reflect meaning.
- Remove unused storage item.
- Introduce `OnMinted` trait.
- Introduce new skeleton treasury module.

### CODE REVIEW

File: substrate/runtime/example/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+
+/// Exported Event type that's generic over the configuration trait.
+// NOTE: External macro-fu expects this type to exist and be generic over
+// the configuration trait.
+pub type Event<T> = RawEvent<
+	<T as balances::Trait>::Balance,
+>;
+
+/// An event in this module. Events are simple means of reporting specific conditions and
+/// cursumstances that have happened that users, Dapps and/or chain explorers would find
```
Comment (dvdplm) [inline substrate/runtime/example/src/lib.rs]: typo: curcumstances

File: substrate/runtime/example/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+/// interested and otherwise difficult to detect.
```
Comment (dvdplm) [inline substrate/runtime/example/src/lib.rs]: s/interested/interesting/

File: substrate/runtime/example/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+		//   - `Foo::kill(); Foo::get()` panics.
+		// - `Foo: default u32`:
+		//   - `Foo::put(1); Foo::get()` returns `1`;
+		//   - `Foo::kill(); Foo::get()` returns `0` (u32::default()).
+		// e.g. Foo: u32;
+		// e.g. pub Bar get(bar): default map [ T::AccountId => Vec<(T::Balance, u64)> ];
+		//
+		// For basic value items, you'll get a type which implements
+		// `runtime_support::StorageValue`. For map items, you'll get a type which
+		// implements `runtime_support::StorageMap`.
```
Comment (dvdplm) [inline substrate/runtime/example/src/lib.rs]: Is it worth mentioning that `StorageMap` is `HashMap<Vec<u8>, Vec<u8>>`?

File: substrate/runtime/example/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+
+// The main implementation block for the module. Functions here fall into three broad
+// categories:
+// - Implementations of dispatch functions. The dispatch code generated by the module macro
+// expects each of its functions to be implemented.
+// - Public interface. These are functions that are `pub` and generally fall into inspector
+// functions that do not write to storage and operation functions that do.
+// - Private functions. These are your usual private utilities unavailable to other modules.
+impl<T: Trait> Module<T> {
+	/// Deposit one of this module's events. This function doesn't change.
```
Comment (dvdplm) [inline substrate/runtime/example/src/lib.rs]: Not sure I understand the `This function doesn't change.` part. Can you rephrase?

File: substrate/runtime/example/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+	//
+	// Since this is a dispatched function there are two extremely important things to
+	// remember:
+	//
+	// - MUST NOT PANIC: Under no circumstances (save, perhaps, storage getting into an
+	// irreparably damaged state) must this function panic.
+	// - NO SIDE-EFFECTS ON ERROR: This function must either complete totally (and return 
+	// `Ok(())` or it must have no side-effects on storage and return `Err('Some reason')`.
+	//
+	// The first is relatively easy to audit for - just ensure all panickers are emoved from
```
Comment (dvdplm) [inline substrate/runtime/example/src/lib.rs]: typo: emoved

File: substrate/runtime/example/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+	// logic that executes in production (which you do anyway, right?!). To ensure the second 
+	// is followed, you should do all tests for validity at the top of your function. This
+	// is stuff like checking the sender (`aux`) or that state is such that the operation
+	// makes sense.
+	// 
+	// Once you've determined that it's all good, then enact the operation and change storage.
+	// If you can't be certain that the operation will succeed without substantial computation
+	// then you have a classic blockchain attack scenario. The normal way of managing this is
+	// to attach a bond to the operation. as the first major alteration of storage, reserve
```
Comment (dvdplm) [inline substrate/runtime/example/src/lib.rs]: s/as/As/

File: substrate/runtime/example/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+
+		// Calculate the new value.
+		let new_dummy = dummy.map_or(increase_by, |dummy| dummy + increase_by);
+
+		// Put the new value into storage.
+		<Dummy<T>>::put(new_dummy);
+		// Will also work with a reference:
+		// <Dummy<T>>::put(&new_dummy);
+
+		// Let's deposit an event to let the outside world this happened.
```
Comment (dvdplm) [inline substrate/runtime/example/src/lib.rs]: s/…outside world this…/…outside world know this…/

File: substrate/runtime/example/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+
+		// All good.
+		Ok(())
+	}
+
+	// Implementation of a priviledged call. This doesn't have an `aux` parameter because
+	// it's not (directly) from an extrinsic, but rather the system as a whole has decided
+	// to execute it. Different runtimes have different reasons for allow priviledged
+	// calls to be executed - we don't need to care why. Because it's priviledged, we can
+	// assume it's a one-off operation and substantial processing/storage/memort can be used
```
Comment (dvdplm) [inline substrate/runtime/example/src/lib.rs]: s/memort/memory/

File: substrate/runtime/example/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+	// without worrying about gamability or attack scenarios.
```
Comment (dvdplm) [inline substrate/runtime/example/src/lib.rs]: s/gamability/gameability/

File: substrate/runtime/example/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+	fn set_dummy(new_value: T::Balance) -> Result {
+		// Put the new value into storage.
+		<Dummy<T>>::put(new_value);
+
+		// All good.
+		Ok(())
+	}
+}
+
+// The trait expresses what should happen when the block is finalised.
```
Comment (dvdplm) [inline substrate/runtime/example/src/lib.rs]: s/The/This/

File: substrate/runtime/example/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+		<Dummy<T>>::kill();
+	}
+}
+
+#[cfg(feature = "std")]
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[serde(deny_unknown_fields)]
+/// The genesis block configuration type. This is a simple default-capable struct that
+/// contains any fields with which this module can be configured at genesis time.
```
Comment (dvdplm) [inline substrate/runtime/example/src/lib.rs]: For my education, what is "genesis time" here? Is it the initial state of the contract? So unrelated to the "block chain genesis" right?

File: substrate/runtime/treasury/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,555 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
```
Comment (dvdplm) [inline substrate/runtime/treasury/src/lib.rs]: s/2017/2018/

File: substrate/runtime/treasury/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,555 @@
...
+extern crate substrate_runtime_balances as balances;
+
+use rstd::ops::{Mul, Div};
+use runtime_support::{StorageValue, StorageMap};
+use runtime_support::dispatch::Result;
+use runtime_primitives::traits::{As, OnFinalise, Zero, RefInto};
+use balances::OnMinted;
+
+/// Our module's configuration trait. All our types and consts go in here. If the
+/// module is dependent on specfiic other modules, then their configuration traits
```
Comment (dvdplm) [inline substrate/runtime/treasury/src/lib.rs]: s/specfiic/specific/

File: substrate/runtime/treasury/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,555 @@
...
+
+// The module declaration. This states the entry points that we handle. The
+// macro looks after the marshalling of arguments and dispatch.
+decl_module! {
+	// Simple declaration of the `Module` type. Lets the macro know what its working on.
+	pub struct Module<T: Trait>;
+
+	#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+	pub enum Call where aux: T::PublicAux {
+		// Put forward a suggestion for spending. A bond of
```
Comment (dvdplm) [inline substrate/runtime/treasury/src/lib.rs]: This looks weird. Rephrase?

File: substrate/runtime/treasury/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,555 @@
...
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Treasury {
+		// Config...
+
+		// Proportion of funds that should be bonded in order to place a proposal. An accepted
+		// proposal gets these back. A rejected proposal doesn't.
+		ProposalBond get(proposal_bond): required Permill;
+		
+		// Proportion of funds that should be bonded in order to place a proposal. An accepted
```
Comment (dvdplm) [inline substrate/runtime/treasury/src/lib.rs]: Is this really a proportion? If so, shouldn't it be a `Permill`?

File: substrate/runtime/treasury/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,555 @@
...
+		// Total funds available to this module for spending.
+		Pot get(pot): default T::Balance;
+
+		// Number of proposals that have been made.
+		ProposalCount get(proposal_count): default ProposalIndex;
+
+		// Proposals that have been made.
+		Proposals get(proposals): map [ ProposalIndex => Proposal<T::AccountId, T::Balance> ];
+
+		// Proposals that have been made.
```
Comment (dvdplm) [inline substrate/runtime/treasury/src/lib.rs]: s/have been made/have been approved/?

File: demo/runtime/src/lib.rs [hunk 17]
```diff
@@ -65,8 +65,8 @@ pub use runtime_primitives::BuildStorage;
 // Workaround for https://github.com/rust-lang/rust/issues/26925 . Remove when sorted.
 #[derive(Clone, Copy, PartialEq, Eq)]
 #[cfg_attr(feature = "std", derive(Debug, Serialize, Deserialize))]
-/// Concrete runtime type used to parameterize the various modules.
-pub struct Concrete;
+/// Runtime runtime type used to parameterize the various modules.
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply demo/runtime/src/lib.rs]: I like short, bold and unequivocal naming. The struct that is created is best described as the Runtime as all runtime logic hangs from it. `ConcreteRuntime` suggests there's an `AbstractRuntime` trait somewhere, which there isn't.

File: substrate/runtime/contract/src/lib.rs [hunk 18]
```diff
@@ -271,8 +271,8 @@ impl<T: Trait> balances::OnFreeBalanceZero<T::AccountId> for Module<T> {
 }
 
 /// Finalization hook for the smart-contract module.
-impl<T: Trait> Executable for Module<T> {
-	fn execute() {
+impl<T: Trait> OnFinalise<T::BlockNumber> for Module<T> {
+	fn on_finalise(_n: T::BlockNumber) {
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply substrate/runtime/contract/src/lib.rs]: feel free to make a PR, but that's out of scope for this one.

File: substrate/runtime/example/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,385 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply substrate/runtime/example/src/lib.rs]: out of scope - this is blanket legal boilerplate and duplicated across the codebase and as such should be handled in its own PR for a blanket upgrade.

File: substrate/runtime/example/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,385 @@
...
+	#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+	pub enum Call where aux: T::PublicAux {
+		// This is just a simple example of how to interact with the module from the external
+		// world.
+		fn accumulate_dummy(aux, increase_by: T::Balance) -> Result = 0;
+	}
+
+	// The priviledged entry points. These are provided to allow any governance modules in 
+	// the runtime to be able to execute common functions. Unlike for `Call` there is no
+	// auxilliary data to encode the sender (since there is no sender). Though still important
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply substrate/runtime/example/src/lib.rs]: happy for the to be an editorial PR on top of this one, to discuss exact wording, but don't consider it important to muck around at this stage.

File: substrate/runtime/example/src/lib.rs (hunk 4 above)
Comment (gavofyork) [reply substrate/runtime/example/src/lib.rs]: removed as superfluous

File: substrate/runtime/example/src/lib.rs (hunk 11 above)
Comment (gavofyork) [reply substrate/runtime/example/src/lib.rs]: The initial state of the blockchain, which dictates this module's initial storage items' values.

File: substrate/runtime/treasury/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,556 @@
...
+		Ok(())
+	}
+
+	fn configure(
+		proposal_bond: Permill,
+		proposal_bond_minimum: T::Balance,
+		spend_period: T::BlockNumber,
+		burn: Permill
+	) -> Result {
+		<ProposalBond<T>>::put(proposal_bond);
```
Comment (pepyakin) [inline substrate/runtime/treasury/src/lib.rs]: This is subtle: we must make sure that this isn't called if we already have any proposals, right? Does this worth a check and/or a comment?

File: substrate/runtime/primitives/src/traits.rs [hunk 22]
```diff
@@ -184,18 +180,18 @@ impl<T:
...
 > SimpleBitOps for T {}
 
-/// Something that can be executed.
-pub trait Executable {
-	fn execute();
+/// The block finalisation trait. Implementing this lets you express what should happen
+/// for your module when the block is ending.
+pub trait OnFinalise<BlockNumber> {
+	/// The block is being finalised. Implement to have something happen.
+	fn on_finalise(_n: BlockNumber) {}
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply substrate/runtime/primitives/src/traits.rs]: it's available in the caller and it's useful in many of the callees.

File: polkadot/parachain/test-chains/basic_add/Cargo.lock [hunk 23]
```diff
@@ -0,0 +1,224 @@
+[[package]]
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply polkadot/parachain/test-chains/basic_add/Cargo.lock]: i think so - it's an entirely separate project

File: substrate/runtime/treasury/src/lib.rs (hunk 21 above)
Comment (gavofyork) [reply substrate/runtime/treasury/src/lib.rs]: Good catch. Needs further work but will just make it safe for now.

**Comment by gavofyork** [comment]
@pepyakin Grumbles addressed

File: substrate/runtime/primitives/src/traits.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (pepyakin) [reply substrate/runtime/primitives/src/traits.rs]: True, but i worry a bit that making a slight improvement two modules we might get a sligh loss in tests: e.g. if we want to test `on_finalise` in some module and that module doesn't care about block number then we will still have to supply it.
